File: src/4_ml_model_bot.py
from time import time
from common.classes.classes import ResponseException
from common.methods.getResearchPapers import getResearchPapers
from common.methods.parseUpdate import UpdateInfo
from common.methods.sendTelegramMessage import sendTelegramMessage
from common.methods.startServer import startServerPolling
from common.methods.summarizeWithML import summarizeWithML

def parse_response(update_info: UpdateInfo):
  start = time()
  sendTelegramMessage(update_info['chat_id'], "Elaborating request...")

  keyword = update_info['message']
  core_ac_response = getResearchPapers(keyword)
  
  abstracts_text = ""
  try:
    if len(core_ac_response['results'])==0:
      #if no paper is found, tell user what happened
      sendTelegramMessage(update_info['chat_id'], 'CoreAC was not able to find any result for '+keyword)
      return
    else:
      #retrive abstract from CoreAc response
      for s in core_ac_response['results']:
        abstracts_text += f"{s['abstract']}\n"
  except:
    print(f"ERROR: {core_ac_response}")
    if('error' in core_ac_response.keys()):
      return_msg = keyword+': Error getting the research papers: '+core_ac_response['error']
    else: 
      return_msg = keyword+': Unknown error getting the research papers. Please retry later.'; 
    sendTelegramMessage(update_info['chat_id'], return_msg)
    return

  core_ac_log = f'CoreAc responded with {len(core_ac_response["results"])} results, out of {core_ac_response["totalHits"]}'
  sendTelegramMessage(update_info['chat_id'], core_ac_log)
  print(core_ac_log)
  print(f"lenght of the abstract composition: {len(abstracts_text)}")

  # Send a message to the user informing him that the summarization could take time
  return_msg = f"Summarizing the paper Abstracts, this can take some time..."
  sendTelegramMessage(update_info['chat_id'], return_msg)

  # call summarizeWithML giving in input the abstracts text and
  # properly handle the ResponseException it could arise, sending a
  # message to the user informing him about the problem, then return
  try:
    hugging_face_obj = summarizeWithML(abstracts_text)
    print('HuggingFace responded')
  except ResponseException:
    return_msg = f"Sorry, request failed at HuggingFace API. Reason: Model is loading, wait some time and retry"
    sendTelegramMessage(update_info['chat_id'], return_msg)
    return

  # Send a message to the user, containing the summarized text.
  # Remember to properly handle a possible exception, if that happen
  # you can access the error message by doing hugging_face_obj['error']
  # send a message informing the user.
  # hint: you can access the summarized text by doing
  # hugging_face_obj[0]['summary_text']

  try:
    return_msg = hugging_face_obj[0]['summary_text']
    sendTelegramMessage(update_info['chat_id'], return_msg)
  except:
    error_msg = hugging_face_obj['error']
    print(f"crashed hugging face api. Reason: {error_msg}")
    return_msg = f"Sorry, request failed at HuggingFace API. Reason: {error_msg}. Retry"
    sendTelegramMessage(update_info['chat_id'], return_msg)
    return

  end = time()
  print(f"request served in {end-start} s")

startServerPolling(parse_response)

File: src/typescript/4_ml_model_bot.ts
import {
  CoreACSearchResponse,
  GetUpdatesResponse,
  HuggingFaceResponse,
  Update,
  UpdateInfo,
} from "./types";
import fetch from "node-fetch";
import { config } from "dotenv";
import { parse } from "path";
config({ path: __dirname + "/./../../.env" });

// retrieve tokens from .env file
const BOT_TOKEN = process.env.BOT_TOKEN;
const CORE_AC_TOKEN = process.env.CORE_AC_TOKEN;
const HUGGING_FACE_TOKEN = process.env.HUGGING_FACE_TOKEN;

// urls
const telegram_url = `https://api.telegram.org/bot${BOT_TOKEN}`;
const core_ac_url = "https://api.core.ac.uk/v3/search/works/";
const huggin_face_url =
  "https://api-inference.huggingface.co/models/google/bigbird-pegasus-large-pubmed";

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function parseUpdate(update: Update) {
  const chat_id = update.message.chat.id.toString();
  const sender = update.message.chat.username;
  const message = update.message.text;

  console.log(`\n${sender} says ${message}`);

  const update_info: UpdateInfo = {
    chat_id,
    sender,
    message,
  };
  return update_info;
}

async function startServerPolling(
  parse_response: (update_info: UpdateInfo) => void
) {
  console.log("Server online. Waiting...");
  // id of the last parsed message
  let last_update = 0;
  while (true) {
    const r = await fetch(`${telegram_url}/getUpdates?offset=${last_update}`);
    const response: GetUpdatesResponse = await r.json();

    if (response.result.length > 0) {
      for (const update of response.result) {
        const update_info = parseUpdate(update);
        parse_response(update_info);
        last_update = update.update_id + 1;
      }
    }
    await sleep(1000);
  }
}

async function getResearchPapers(
  update_info: UpdateInfo,
  CORE_AC_TOKEN: string
) {
  const limit = 5;
  const r = await fetch(
    `${core_ac_url}?q=${update_info.message}&limit=${limit}`,
    {
      headers: { Authorization: `Bearer ${CORE_AC_TOKEN}` },
    }
  );
  const core_ac_response: CoreACSearchResponse = await r.json();
  return core_ac_response;
}

async function sendTelegramMessage(chat_id: string, return_msg: string) {
  await fetch(`${telegram_url}/sendMessage`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      chat_id: chat_id,
      text: return_msg,
    }),
  });
}

async function parse_response(update_info: UpdateInfo) {
  console.time('process request');

  const core_ac_response = await getResearchPapers(update_info, CORE_AC_TOKEN);
  let abstracts_text = "";
  try {
    for (const s of core_ac_response.results) {
      abstracts_text += `${s.abstract}\n`;
    }
  } catch (e) {
    const error_msg: string = core_ac_response["message"];
    console.log(`crashed core ac api. Reason: ${error_msg}`);
    const return_msg = `Sorry, request failed at CoreAc API. Reason: ${error_msg}. Retry`;
    await sendTelegramMessage(update_info.chat_id, return_msg);
    return;
  }

  console.log(
    `CoreAc responded with ${core_ac_response.results.length} results, out of ${core_ac_response.totalHits}`
  );
  console.log(`lenght of the abstract composition: ${abstracts_text.length}`);

  const hugging_face_response = await fetch(huggin_face_url, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${HUGGING_FACE_TOKEN}`,
      "Content-Type": "application/json",
    },
    body: abstracts_text,
  });
  const hugging_face_obj: HuggingFaceResponse =
    await hugging_face_response.json();
  console.log("HuggingFace responded");

  try {
    const return_msg = hugging_face_obj[0].summary_text;
    await sendTelegramMessage(update_info.chat_id, return_msg);
  } catch (e) {
    const error_msg: string = hugging_face_obj["error"];
    console.log(`crashed hugging face api. Reason: ${error_msg}`);
    const return_msg = `Sorry, request failed at HuggingFace API. Reason: ${error_msg}. Retry`;
    await sendTelegramMessage(update_info.chat_id, return_msg);
    return;
  }

  console.timeEnd('process request');
}

startServerPolling(parse_response);
